File: sync.py
#!/usr/bin/env python

import difflib
import hashlib
import os
import shutil
import sys

NORMAL      = '\033[0m'
BOLD        = '\033[1m'
RED         = '\033[31m'
GREEN       = '\033[32m'
BOLD_RED    = '\033[1;31m'
BOLD_GREEN  = '\033[1;32m'

EXCLUDE = ['README', os.path.basename(sys.argv[0])]
home = os.getenv('HOME')

def green(s):
    return GREEN + s + NORMAL

def red(s):
    return RED + s + NORMAL

def bold_green(s):
    return BOLD_GREEN + s + NORMAL

def bold_red(s):
    return BOLD_RED + s + NORMAL

def bold(s):
    return BOLD + s + NORMAL

def format_diff(l):
    l = l.rstrip()
    if l.startswith('---'):
        return bold_red(l)
    elif l.startswith('+++'):
        return bold_green(l)
    elif l.startswith('-'):
        return red(l)
    elif l.startswith('+'):
        return green(l)
    else:
        return l

for root, dirs, files in os.walk('.'):
    if root == '.':
        files = list(set(files) - set(EXCLUDE))
    for d in dirs:
        if d.startswith('.'):
            dirs.remove(d)
    for f in files:
        if f.startswith('.'):
            continue
        p1 = os.path.join(root, f)
        if root == '.':
            p2 = os.path.join(home, '.' + f)
        else:
            p2 = os.path.join(home, '.' + root[2:], f)
        if os.path.exists(p2):
            h2 = hashlib.sha1(open(p2).read()).digest()
        else:
            print bold('\n==> %s does not exist' % p2)
            ans = raw_input('Install current version ? [Y|n] ')
            if ans.lower() in [ '', 'y' ]:
                dir = os.path.dirname(p2)
                if not os.path.exists(dir):
                    os.makedirs(dir)
                shutil.copyfile(p1, p2)
            continue
        h1 = hashlib.sha1(open(p1).read()).digest()
        if h1 == h2:
            continue

        print bold('\n==> %s and %s are different' % \
            (p1, p2.replace(home, '~')))

        ans = 'd'
        while ans == 'd':
            while True:
                print '\nYou can:'
                print '    k) Keep local version without modifying VCS version'
                print '    i) Install version in version control'
                print '    u) Use local version and copy it in VCS'
                print '    d) Show diff'
                ans = raw_input('Your choice: [K|i|u|d] ')
                ans = ans.lower()
                if ans in ('', 'k', 'i', 'u', 'd'):
                    break
        
            if ans == 'i':
                shutil.copyfile(p1, p2)
            elif ans == 'u':
                shutil.copyfile(p2, p1)
            elif ans == 'd':
                d = difflib.unified_diff(open(p2, 'rt').readlines(),
                                         open(p1, 'rt').readlines(),
                                         p2.replace(home, '~'),
                                         p1)
                print '\n'.join([format_diff(l) for l in d])

# vim:et:sw=4:ts=4:

File: sync.go
package main

import (
	"bufio"
	"bytes"
	"crypto/md5"
	"flag"
	"fmt"
	"io"
	"log"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

var excluded = map[string]struct{}{
	"README.md": {},
	"sync.go":   {},
	"sync.py":   {},
}

var install = flag.Bool("install", false, "install new or modified files without asking question")

func bold(s string) string {
	return "\033[1m" + s + "\033[0m"
}

func compareFiles(path1, path2 string) (bool, error) {
	md5s := [2][]byte{}

	for i, fn := range [2]string{path1, path2} {
		f, err := os.Open(fn)
		if err != nil {
			return false, fmt.Errorf("opening %s: %s", fn, err)
		}

		h := md5.New()
		if _, err := io.Copy(h, f); err != nil {
			return false, fmt.Errorf("hashing %s: %s", fn, err)
		}

		md5s[i] = h.Sum(nil)

		f.Close()
	}

	return bytes.Equal(md5s[0], md5s[1]), nil
}

func ask(text string) string {
	reader := bufio.NewReader(os.Stdin)
	fmt.Print(text)
	ans, _ := reader.ReadString('\n')
	return strings.TrimSpace(ans)
}

func copyFile(src, dst string) error {
	in, err := os.Open(src)
	if err != nil {
		return err
	}
	defer in.Close()

	out, err := os.Create(dst)
	if err != nil {
		return err
	}
	defer func() {
		cerr := out.Close()
		if err == nil {
			err = cerr
		}
	}()

	if _, err = io.Copy(out, in); err != nil {
		return err
	}
	err = out.Sync()

	return err
}

func installFile(src, dst string) error {
	fmt.Printf("Installing %s => %s\n", src, dst)
	if err := os.MkdirAll(filepath.Dir(dst), 0755); err != nil {
		return err
	}
	return copyFile(src, dst)
}

func showDiff(src, dst string) {
	out, _ := exec.Command("diff", "-u", "--ignore-blank-lines", "--color=always", src, dst).Output()
	fmt.Printf("%s", out)
}

func installNewFile(src, dst string) error {
	fmt.Printf(bold("\n==> %s does not exist\n"), dst)

	if *install {
		return installFile(src, dst)
	}

	for {
		c := ask("Install current version? [Y|n|q] ")
		switch {
		case c == "", c == "y", c == "Y":
			return installFile(src, dst)
		case c == "n", c == "N":
			return nil
		case c == "q", c == "Q":
			os.Exit(0)
		}
	}
}

func handleDifferentFiles(src, dst string) error {
	fmt.Printf(bold("\n==> %s and %s are different\n"), src, dst)

	if *install {
		return installFile(src, dst)
	}

	for {
		fmt.Println("You can:")
		fmt.Println("    d) Show difference between files")
		fmt.Println("    i) Install version from this repository")
		fmt.Println("    k) Keep local version")
		fmt.Println("    u) Use local version and copy it in this repository")
		fmt.Println("    q) Quit")
		c := ask("Your choice: [d|i|K|u|q] ")
		switch {
		case c == "d", c == "D":
			showDiff(dst, src)
		case c == "i", c == "I":
			return installFile(src, dst)
		case c == "", c == "k", c == "K":
			return nil
		case c == "u", c == "U":
			return installFile(dst, src)
		case c == "q", c == "Q":
			os.Exit(0)
		}
	}
}

func analyzeFile(path string, info os.FileInfo, err error) error {
	switch {
	case info.IsDir() || path[0] == '.':
		return nil
	case err != nil:
		fmt.Printf("ERROR: reading %s: %s\n", path, err)
		return nil
	}

	if _, ok := excluded[filepath.Base(path)]; ok {
		return nil
	}

	realpath := filepath.Join(os.Getenv("HOME"), "."+path)
	if _, err := os.Stat(realpath); os.IsNotExist(err) {
		if err := installNewFile(path, realpath); err != nil {
			fmt.Printf("ERROR: %s\n", err)
		}
		return nil
	}

	same, err := compareFiles(path, realpath)
	switch {
	case err != nil:
		fmt.Printf("ERROR: %s", err)
		return nil
	case same:
		return nil
	}

	if err := handleDifferentFiles(path, realpath); err != nil {
		fmt.Printf("ERROR: %s\n", err)
	}

	return nil
}

func main() {
	flag.Parse()

	err := filepath.Walk(".", analyzeFile)
	if err != nil {
		log.Fatalf("ERROR: %s\n", err)
	}
}
